java/vector/src/main/java/org/apache/arrow/vector/stream/ArrowStreamReader.java

```diff
@@ -0,0 +1,180 @@
+  }
+
+  /**
+   * Returns the total number of batches in the stream. Returns -1 if this is unknown.
+   */
+  public int getTotalBatches() {
+    Preconditions.checkState(this.header != null, "Must call init() first.");
+    return header.getTotalBatches();
+  }
+
+  /**
+   * Returns the schema for all records in this stream.
+   */
+  public Schema getSchema () {
+    Preconditions.checkState(this.header != null, "Must call init() first.");
+    return header.getSchema();
+  }
+
+  public long bytesRead() { return bytesRead; }
+
+  /**
+   * Reads and returns the next ArrowRecordBatch. Returns null if this is the end
+   * of stream.
+   */
+  public ArrowRecordBatch nextRecordBatch() throws IOException {
+    Preconditions.checkState(this.in != null, "Cannot call after close()");
+    Preconditions.checkState(this.header != null, "Must call init() first.");
+
+    ByteBuffer intBuffer = ByteBuffer.allocate(4);
+    int numRead = readFully(intBuffer);
+    if (numRead == 0) {
+      // No more bytes in stream. No more batches.
+      return null;
+    }
+    if (numRead != intBuffer.capacity()) {
+      throw new InvalidArrowStreamException(
+          "Unexpected end of stream reading batch metadata length.");
+    }
+    int metadataLength = ArrowReadUtil.bytesToInt(intBuffer.array());
+    intBuffer.rewind();
+    if (readFully(intBuffer) != intBuffer.capacity()) {
+      throw new InvalidArrowStreamException(
+          "Unexpected end of stream reading batch body length.");
+    }
+    int bodyLength = ArrowReadUtil.bytesToInt(intBuffer.array());
+    Preconditions.checkState(metadataLength >= 0);
+    Preconditions.checkState(bodyLength >= 0);
+    int dataLen = metadataLength + bodyLength;
+
+    final ArrowBuf buffer = allocator.buffer(dataLen);
+    LOGGER.debug("allocated buffer " + buffer);
+    if (readFully(buffer, dataLen) != dataLen) {
+      throw new InvalidArrowStreamException(
+          "Unexpected end of stream reading batch bytes: " + dataLen);
+    }
+    return ArrowReadUtil.constructRecordBatch(buffer, metadataLength, bodyLength);
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (this.in != null) {
+      in.close();
+      in = null;
+    }
+  }
+
+  /**
+   * Reads bytes into buffer until it is full (buffer.remaining() == 0). Returns the
+   * number of bytes read which can be less than full if there are no more.
+   */
+  private int readFully(ByteBuffer buffer) throws IOException {
+    int totalRead = 0;
+    while (buffer.remaining() != 0) {
+      int read = in.read(buffer);
+      if (read < 0) return totalRead;
+      totalRead += read;
+      if (read == 0) break;
+    }
+    this.bytesRead += totalRead;
+    return totalRead;
+  }
+
+  /**
+   * Reads up to len into buffer. Returns bytes read.
+   */
+  private int readFully(ArrowBuf buffer, int l) throws IOException {
+    int n = readFully(buffer.nioBuffer(buffer.writerIndex(), l));
+    buffer.writerIndex(n);
+    return n;
+  }
```

Comment:
Yea, there is too much code duplication now. I moved a bit of it around but there is more to do. I was planning to do this in a follow up if this patch makes sense.